Tests/UpstreamIntegrationTests/UpstreamIntegrationTests.swift

```diff
@@ -0,0 +1,213 @@
+class UpstreamIntegrationTests: XCTestCase {
+    private var edgeEnvironment: EdgeEnvironment = .prod
+    private var edgeLocationHint: EdgeLocationHint? = nil
+    
+    private let testingDelegate = NetworkTestingDelegate()
+    
+    let LOG_SOURCE = "SampleFunctionalTests"
+    
+    let asyncTimeout: TimeInterval = 10
+
+    override func setUp() {
+        let networkService = IntegrationTestNetworkService()
+        networkService.testingDelegate = testingDelegate
+        ServiceProvider.shared.networkService = networkService
+        
+        // Extract Edge Network environment level from shell environment
+        if let environment = EdgeEnvironment() {
+            self.edgeEnvironment = environment
+        }
+        print("Using Edge Network environment: \(edgeEnvironment.rawValue)")
+        // Extract Edge location hint from shell environment
+        if let locationHint = EdgeLocationHint() {
+            self.edgeLocationHint = locationHint
+        }
+        
+        let waitForRegistration = CountDownLatch(1)
+        MobileCore.setLogLevel(.trace)
+        MobileCore.configureWith(appId: "94f571f308d5/6b1be84da76a/launch-023a1b64f561-development")
+        MobileCore.registerExtensions([Identity.self, Edge.self], {
+            print("Extensions registration is complete")
+            waitForRegistration.countDown()
+        })
+        XCTAssertEqual(DispatchTimeoutResult.success, waitForRegistration.await(timeout: 2))
+        
+        // Set Edge location hint value if one is set for the test
+        if edgeLocationHint != nil {
+            print("Setting Edge location hint to: \(String(describing: edgeLocationHint?.rawValue))")
+            Edge.setLocationHint(edgeLocationHint?.rawValue)
+        }
+        else {
+            print("No preset Edge location hint is being used for this test.")
+        }
+    }
+    
+    public override func tearDown() {
+        super.tearDown()
+
+        // to revisit when AMSDK-10169 is available
+        // wait .2 seconds in case there are unexpected events that were in the dispatch process during cleanup
+        usleep(200000)
+        EventHub.reset()
+        UserDefaults.clearAll()
+        FileManager.default.clearCache()
+    }
+    
+    // TODO: create specific listeners for type: Edge + source: * (wildcard) and capture all the response handles for the test event
+    // TODO: create specific listeners for error responses
+    func registerEdgeLocationHintListener(_ listener: @escaping EventListener) {
+        MobileCore.registerEventListener(type: FunctionalTestConst.EventType.EDGE, source: "locationHint:result", listener: listener)
+    }
+    
+    /// Converts a JSON string into the provided type.
+    ///
+    /// NOTE: caller is reponsible for providing the correct casting type resulting JSON, otherwise decoding will fail
+    func convertToJSON<T>(_ jsonString: String) -> T? {
```

Comment:
I suggest moving private helper functions at the end of the file so you can easily follow the flow of the test file:
properties

// main test class body
setup
tearDown
test cases

helpers